Add AZURE_LOG() to log plugin runtime information

coko156 opened the pull request:
# Logs  for mysqlnd_azure plugin

This patch is added for print some running information of mysqlnd_azure to a local file.
Current status: Under developing.

## New Added Configuration
### `mysqlnd_azure.logfilePath`

Name | mysqlnd\_azure.logfilePath
:----- | :------
Comment | Filename that you want your log writes to.
Type | String
Accepted Value | String length <= 255, legal filename string
Default | mysqlnd\_azure\_runtime.log
Dynamic | NO
Note | 1. This variable was inited at PHP Module Init (MINIT) period, and immutable at runtime. a call of `ini_set('mysqlnd_azure.logLevel', '$VAL'); ` would fail. <br> 2. Filename length cannot exceed the file system's restriction(255), otherwise system will use the default filename and throw a warning "Given logfile name too long, redirected to default: mysqlnd_azure_runtime.log"

### `mysqlnd_azure.logLevel`

Name | mysqlnd\_azure.logLevel
:----- | :------
Comment | The verbose level of log we generate.
Type | Enumeration
Accepted Value | [ 0 \| 1 \| 2 \| 3 ]
Default | 0 (OFF)
Dynamic | Yes
Note | 1. When the current `mysqlnd_azure.logLevel > 0`, change logLevel at runtime will be logged.


#### Loglevel supported
we support 3 level of logs for user:
- [DEBUG]:
  1. source code path(inside which function);
  2. connection information(except password);
  3. verbose running status (e.g. customer given redirected server information directly... )
- [INFO]:
  1. connection fallback when mysqlnd\_azure.enableRedirection=preferred with details (cannot get redirection
      info from OK packet/ try connect using redirect information failed , ...);
  2. cache operations
- [ERROR]:
  1. any operation may cause customer see connection break (e.x. original conn/ redirected\_conn
    established failed, ... and connection fall back when mysqlnd\_azure.enableRedirection = ON)


> [SYSTM] maked log rows for Azure mysqlnd Log Module itself, when current
> `mysqlnd_azure.logLevel > 0`, any operation related to log system itself will be logged.
> For example, if someone call  `ini_set('mysqlnd_azure.logLevel', '2');` at runtime, there
> may appear a log like 
> ```
> 2020-03-27 09:09:05 [SYSTM] mysqlnd_azure.logLevel changed: 3 -> 2
> ```

#### Level value relationship to level logged
- 0: No log written
- 1: [SYSTM] + [ERROR]
- 2: [SYSTM] + [ERROR] + [INFO]
- 3: [SYSTM] + [ERROR] + [INFO] + [DEBUG]


### Configuration Example:

> You can add to section [mysqlnd\_azure] in file `php.ini`

```
[mysqlnd_azure]
mysqlnd_azure.enableRedirect = 2
mysqlnd_azure.logfilePath = "test.log"
mysqlnd_azure.logLevel = 3
```

### phpinfo() / php -i

After successfully configured, try `php -i`, there will be var:val pairs listed below
mysqlnd\_azure section, like:

```
mysqlnd_azure

mysqlnd_azure => enableRedirect
enableRedirect => preferred
logfilePath => santotest.log
logLevel => 3
```

## What the logs would be (example)
After turned on the option mentioned before, you will see something like :
```
2020-03-13 11:00:13 [ERROR] Connection cannot established with given information, please check the server availability, or your parameters given
 2020-03-13 11:00:54 [DEBUG] mysqlnd_azure.c: mysqlnd_azure::connect()
 2020-03-13 11:00:54 [DEBUG] mysqlnd_azure.enableRedirect: PREFERRD
 2020-03-13 11:00:54 [INFO ] SSL not enabled, and mysqlnd_zaure.enableRedirect if PREFERRED, connection will go through gateway.
 2020-03-13 11:00:54 [DEBUG] mysqlnd_azure.c: mysqlnd_azure::connect()
 2020-03-13 11:00:54 [INFO ] No cache found
 2020-03-13 11:00:54 [DBEUG] mysqlnd_azure.c: mysqlnd_azure_data::connect()
 2020-03-13 11:00:54 [DEBUG] Connection Information: host=127.0.0.1 user=santo db=test port=3306 flags=131072 persistent=1 state=0
 2020-03-13 11:00:54 [ERROR] Connection cannot established with given information, please check the server availability, or your parameters given
 2020-03-13 11:00:54 [DEBUG] mysqlnd_azure.c: mysqlnd_azure::connect()
 2020-03-13 11:00:54 [INFO ] No cache found
 2020-03-13 11:00:54 [DBEUG] mysqlnd_azure.c: mysqlnd_azure_data::connect()
 2020-03-13 11:00:54 [DEBUG] Connection Information: host=santo.mysql.database.azure.com user=test@santo db= port=3306 flags=131072 persistent=1 state=0
 2020-03-13 11:00:54 [DEBUG] Classical connection OK, try to get REDIRECTION information
 2020-03-13 11:00:54 [DEBUG] mysqlnd_azure.c: get_redirect_info()
 2020-03-13 11:00:54 [DEBUG] Successfully get redirection information, try to connect with redirection connection
 2020-03-13 11:00:54 [DEBUG] mysqlnd_azure.c: set_redirect_client_options()
 2020-03-13 11:00:55 [DEBUG] Redirect conntion established.
 2020-03-13 11:00:55 [DEBUG] Prepare the context of new connection
 2020-03-13 11:00:55 [DEBUG] mysqlnd_azure.c: mysqlnd_azure::connect()
 2020-03-13 11:00:55 [DEBUG] mysqlnd_azure.enableRedirect: PREFERRD
 2020-03-13 11:00:55 [INFO ] SSL not enabled, and mysqlnd_zaure.enableRedirect if PREFERRED, connection will go through gateway.
```

santo-ms committed:
test

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/signed)](https://cla.opensource.microsoft.com/microsoft/mysqlnd_azure?pullRequest=11) <br/>All CLA requirements met.

dionysus1016 commented on mysqlnd_azure.c:
```diff
@@ -29,16 +29,23 @@
 struct st_mysqlnd_conn_methods org_conn_m;
 struct st_mysqlnd_conn_methods* conn_m;
 
+FILE *logfile;
```
set the value in explicit way. by default it's nullptr, but people don't know whether the default value is valid or not.

dionysus1016 commented on mysqlnd_azure.c:
```diff
@@ -29,16 +29,23 @@
 static enum_func_status
 set_redirect_client_options(MYSQLND_CONN_DATA * const conn, MYSQLND_CONN_DATA * const redirectConn)
 {
+  if (MYSQLND_AZURE_G(logLevel) > 2) {
```
merge these log level checking into log function.

dionysus1016 commented on mysqlnd_azure.c:
```diff
@@ -610,6 +675,15 @@ MYSQLND_METHOD(mysqlnd_azure, connect)(MYSQLND * conn_handle,
+  int logLevel = MYSQLND_AZURE_G(logLevel);
+  if (logLevel > 0) {
+    char *logfilePath = ZSTR_VAL(MYSQLND_AZURE_G(logfilePath));
+    OPEN_LOGFILE(logfilePath);
```
what if we failed to open the log file>

GuuBu commented on mysqlnd_azure.c:
```diff
@@ -340,23 +365,38 @@ MYSQLND_METHOD(mysqlnd_azure_data, connect)(MYSQLND_CONN_DATA ** pconn,
         //Already use redirected connection, or the connection string is a redirected one
         if (strcmp(redirect_host, hostname.s) == 0 && strcmp(redirect_user, username.s) == 0 && ui_redirect_port == port) {
             DBG_ENTER("[redirect]: Is using redirection, or redirection info are equal to origin, no need to redirect");
+            if (MYSQLND_AZURE_G(logLevel) > 2) {
```
Here connection is already established. So maybe "do nothing for redirection." instead of "do nothing and just connect it."?

dionysus1016 reviewed (approved):
so far we will ignore these logs if we failed to open the file, we need to make sure this is expected. Others looks good to me.

GuuBu commented on mysqlnd_azure.c:
```diff
@@ -366,13 +406,19 @@ MYSQLND_METHOD(mysqlnd_azure_data, connect)(MYSQLND_CONN_DATA ** pconn,
             MYSQLND* redirect_conneHandle = mysqlnd_init(MYSQLND_CLIENT_KNOWS_RSET_COPY_DATA, conn->persistent); //init MYSQLND but only need only MYSQLND_CONN_DATA here
             if(!redirect_conneHandle) {
+              if (MYSQLND_AZURE_G(logLevel) > 0) {
+                AZURE_LOG("DEBUG", "Cannot connet with refirection information, redirect connection handshake failed");
```
Here the handshake for redirection is not started yet. This step is to check whether the object is initialized correctly using mysqlnd_init, the possible fail reason can be oom. So may be change to " Init redirect_conneHandle failed" ? Which is same as debug formatted log

GuuBu commented on mysqlnd_azure.c:
```diff
@@ -366,13 +406,19 @@ MYSQLND_METHOD(mysqlnd_azure_data, connect)(MYSQLND_CONN_DATA ** pconn,
                 if(MYSQLND_AZURE_G(enableRedirect) == REDIRECT_ON) {
                     //REDIRECT_ON, abort the original connection and return error
                     conn->m->send_close(conn);
                     SET_CLIENT_ERROR(conn->error_info, MYSQLND_AZURE_ENFORCE_REDIRECT_ERROR_NO, UNKNOWN_SQLSTATE, "Connection aborted because init redirection failed.");
```
I noticed AZURE_LOG will write dupliate log as SET_CLIENT_ERROR step for this error no : MYSQLND_AZURE_ENFORCE_REDIRECT_ERROR_NO. So maybe we can use the identical error message in these cases, with extra message for option choice, and here missing one ERROR log "mysqnd_azure.enableRedirect=on. Connection aborted because init redirection failed."

coko156 replied:
No need to be consistent with `MYSQLND_AZURE_ENFORCE_REDIRECT_ERROR_NO`, unless this number is used by some other system.
Our goal is to know the context "redirectHandle init failed && mysqlnd_azure.enableRedirection value" if some  unexpected behavior happen.
I want to know why must use this ERROR Number here?

coko156 replied:
> I noticed AZURE_LOG will write dupliate log as SET_CLIENT_ERROR step for this error no : MYSQLND_AZURE_ENFORCE_REDIRECT_ERROR_NO. So maybe we can use the identical error message in these cases, with extra message for option choice, and here missing one ERROR log "mysqnd_azure.enableRedirect=on. Connection aborted because init redirection failed."

Seems there is no need to be consistent with the Error Number `MYSQLND_AZURE_ENFORCE_REDIRECT_ERROR_NO`, unless it is used by other systems.

Our goal is to get the context (`mysqlnd_azure.enableRedirect` value, and `redirectHandle ` init fail or success) when unexpected scenarios happen.

If this Error Number cannot offer more information to explain the context, why we should be consistent with `SET_CLIENT_ERROR`?

GuuBu replied:
You are right. I bring this out as I think customer might check both logs, and the inconsistency might cause confusion for the same part. It's all right to me if you decide to skip some or use different message.

GuuBu commented on mysqlnd_azure.c:
```diff
@@ -366,13 +406,19 @@ MYSQLND_METHOD(mysqlnd_azure_data, connect)(MYSQLND_CONN_DATA ** pconn,
                     goto err;
                 } else {
+                  if (MYSQLND_AZURE_G(logLevel) > 1) {
+                    AZURE_LOG("INFO ", "Redirection connection failed, mysqlnd_azure.enableRedirect = PREFERRED, so fall back to classical connection.");
```
Same here, the handshake is not started yet. So maybe we should state the error: " Init redirection object failed" instead of "Redirection connection failed"

GuuBu commented on mysqlnd_azure.c:
```diff
@@ -387,6 +433,9 @@ MYSQLND_METHOD(mysqlnd_azure_data, connect)(MYSQLND_CONN_DATA ** pconn,
             //init redirect_conn options failed, just use the proxy connection
             if (ret == FAIL) {
+              if (MYSQLND_AZURE_G(logLevel) > 0) {
+                AZURE_LOG("ERROR", "Set redirection information to redirect conn failed.");
```
here is to check whether the operation of copying options  from direct conn object to redirect conn object failed. So maybe we can use the same mssage for debug, i.e. "Init redirect_conn options failed"

GuuBu commented on mysqlnd_azure.c:
```diff
@@ -397,6 +446,9 @@ MYSQLND_METHOD(mysqlnd_azure_data, connect)(MYSQLND_CONN_DATA ** pconn,
                     SET_CLIENT_ERROR(conn->error_info, MYSQLND_AZURE_ENFORCE_REDIRECT_ERROR_NO, UNKNOWN_SQLSTATE, "Connection aborted because init redirection failed.");
```
Same here for error message. Here is an error message when mysqlnd_azure.enableRedirect=on

coko156 replied:
Put those logs inside the if/else

GuuBu commented on mysqlnd_azure.c:
```diff
@@ -413,6 +465,10 @@ MYSQLND_METHOD(mysqlnd_azure_data, connect)(MYSQLND_CONN_DATA ** pconn,
             if (redirectState == PASS) { //handshake with redirect_conn succeeded, replace original connection info with redirect_conn and add the redirect info into cache table
+
+              if (MYSQLND_AZURE_G(logLevel) > 2) {
+                AZURE_LOG("DEBUG", "Redirect conntion established.");
```
conntion typo

GuuBu commented on mysqlnd_azure.c:
```diff
@@ -447,6 +506,9 @@ MYSQLND_METHOD(mysqlnd_azure_data, connect)(MYSQLND_CONN_DATA ** pconn,
 
                 if (MYSQLND_AZURE_G(enableRedirect) == REDIRECT_PREFERRED) {
+                  if (MYSQLND_AZURE_G(logLevel > 1)) {
+                    AZURE_LOG("INFO ", "Redirection conn connect failed, mysqlnd_azure.enableRedirect = PREFERRED, fall back to classical connection.");
```
typo. And maybe we can specify the error reason due to handshake here

GuuBu replied:
And for the else cluase, //mysqlnd_azure.enableRedirect=on and redirect connection failed, there is a corresponding error message, the original established connection is closed here, and it will use MySQL's error message as return message. Maybe we should log a ERROR line here to indiate that the original established connection is closed since mysqlnd_azure.enableRedirect=on

coko156 replied:
But you have moved the redirect_conn context to conn after conn resources released.

> Maybe we should log a ERROR line here to indiate that the original established connection is closed since mysqlnd_azure.enableRedirect=on
 Why should we log this?

Besides, I wonder why we copy the redirect_conn context back to conn?

GuuBu replied:
I think a log here may be good since redirect_handshake failed here, and because enableRedirect=on, the original conn is also closed here. That's the reason why no connection is avaible. But since it will finally return redirect_handshake's error, no log here should also be fine.

The reason I copy context back to pconn is that I want to use the error section (below in this method) and the up level to handle the error for me, and error is stored in redirect_conn object, pconn is also a parent object, it will bring the error conn object back to uplevel, and report error correspondly.   Copy back to pfc is actually not necessary in error case since this pointer will only be used below in after_conn section. You may ignore it

GuuBu commented on mysqlnd_azure.c:
```diff
@@ -471,6 +533,9 @@ MYSQLND_METHOD(mysqlnd_azure_data, connect)(MYSQLND_CONN_DATA ** pconn,
 after_conn:
     {
+      if (MYSQLND_AZURE_G(logLevel) > 2) {
+        AZURE_LOG("DEBUG", "Prepare the context of new connection");
```
after_conn will set host/user/xx field to the conn object and run init_commands and then return to up level. Maybe "Post operations after connection establishment" here?

coko156 replied:
> Post operations after connection establishment

What do you mean?

GuuBu replied:
I mean here the handshake already finished, it will only do some post operation like fill the field for the conn object, run init_command. Prepare context looks for me means something before handshake. You may keep it if you think it is just fine.

GuuBu commented on mysqlnd_azure.c:
```diff
@@ -634,12 +711,24 @@ MYSQLND_METHOD(mysqlnd_azure, connect)(MYSQLND * conn_handle,
                     SET_CLIENT_ERROR((*pconn)->error_info, MYSQLND_AZURE_ENFORCE_REDIRECT_ERROR_NO, UNKNOWN_SQLSTATE, "mysqlnd_azure.enableRedirect is on, but SSL option is not set in connection string. Redirection is only possible with SSL.");
                     (*pconn)->m->local_tx_end(*pconn, this_func, FAIL);
                     (*pconn)->m->free_contents(*pconn);
                     return FAIL;
```
Here is also a return. Need to   CLOSE_LOGFILE(); here too

coko156 replied:
Your are right.

But I don't know why use `return FAIL;` here and `DBG_RETURN(ret);` at the end?

GuuBu replied:
Yeah, DEG_RETURN(FAIL) should be better

GuuBu commented on mysqlnd_azure.c:
```diff
@@ -610,6 +675,15 @@ MYSQLND_METHOD(mysqlnd_azure, connect)(MYSQLND * conn_handle,
+  }
+
+  if (logLevel > 2) {
+    AZURE_LOG("DEBUG", "mysqlnd_azure.c: mysqlnd_azure::connect()");
```
mysqlnd_azure.enableRedirect has unique value for same connection phase. So instead of log for every possible option value each time, we may log it only once outside the ifs/elses

coko156 replied:
We print this information only when `mysqlnd_azure.logLevel >= 3` (very verbosely, ERROR, INFO, DEBUG).
If `mysqlnd_azure.logLevel == 1`, we only log ERROR level message(error that trigger a connection break);  when `mysqlnd_azure.logLevel == 2`, ERROR + INFO are logged, INFO means redirect connection fallbacks and cache usage information.

If someone changed this value, we can still find the context when connection broken.

If we only log it once, we cannot specify where the context came from under highly-currenct usage.

GuuBu replied:
Don't really understand you point here.. Maybe there is misunderstanding. What I mean here is only about logLevel 2, those three parts
if (logLevel > 2) {

                    AZURE_LOG("DEBUG", "mysqlnd_azure.enableRedirect: PREFERRD/ON/OFF");
can be stated outside the if/else only once. I understand this as they are in equivalent position in any log level

coko156 replied:
There will be at  most one mysqlnd_azure.enableRedirect value logged per connection, because there is at most one fallback/connection break happen inside a connection. It seems number of logs will no more than we log it at every entry of mysqlnd_azure::connect().

Also, we only need to get that value when unexpected events happen. I dont think we should log the value down every time call mysqlnd_azure::connect().

There is a  huge interval between function entry and exception position., we cannot confirm whether the `enabledRedirection = XX` near a `some exception occurred` belong to a same connection or not.

GuuBu replied:
If one file will log for multi processes, then it is true we cannot make sure. But what I means here is not to only print out once, I just suggest log it for if level2 once outside the if/else, only for code clearance..

GuuBu commented on utils.h:
```diff
@@ -0,0 +1,36 @@
+#ifndef _UTILS_H
```
Maybe name the file as log_utils or log_file_utils for clearance?
And for OPEN_LOGFILE and CLOSE_LOGFILE, instead of each pair in mysqlna_azure::connect. The function pair RINIT/RSHUNDOWN or GINIT/GSHUTDOWN/MSHUTDOWN should be a better place. Since we can make sure it will be called, and if defined in RINT/RSHUTDOWN, it will be opened/closed only once per srcipt run, if defined in GINIT/GSHUTDOWN/MSHUTDOWN, it will be opened/closed only once per thread/process depending on PHP thread mode.

GuuBu commented on php_mysqlnd_azure.c:
```diff
@@ -52,11 +52,31 @@ static ZEND_INI_MH(OnUpdateEnableRedirect)
 /* {{{ PHP_INI */
 PHP_INI_BEGIN()
 STD_PHP_INI_ENTRY("mysqlnd_azure.enableRedirect", "preferred", PHP_INI_ALL, OnUpdateEnableRedirect, enableRedirect, zend_mysqlnd_azure_globals, mysqlnd_azure_globals)
+STD_PHP_INI_ENTRY("mysqlnd_azure.logfilePath", "mysqlnd_azure_runtime.log", PHP_INI_ALL, OnUpdateEnableLogfile, logfilePath, zend_mysqlnd_azure_globals, mysqlnd_azure_globals)
```
It will be great if we can add unit tests to test the option set is correct. You may refer to the ./tests/mysqli_azure_option_test.phpt for reference.

GuuBu replied:
And we may also need to print out these two options when call phpinfo(); or php -i like that for enabeRedirect in PHP_MINFO_FUNCTION

coko156 replied:
resolved the `php -i`

santo-ms committed:
test2

[+] Construct/Destruct inside MINIT/MSHUTDOWN MACROs
[+] php info
And some modifications.

santo-ms committed:
test3

[+] make logfilePath immutable
[+] change some function order in case of crash
[+] [SYSTM] level used by inner debug, not affected by loglevel

santo-ms committed:
test4

[+] Filename length restriction (255)
[+] Expose some logfile error to upper PHP module using php_error_docref()
[+] package.xml placeholder

santo-ms committed:
test5

[+] document
[M] README.md

coko156 committed:
Update mysqlnd_azure_log.md

[m] Fix some typos

santo-ms committed:
test6

[+] `[mysqlnd_azure] tag for error threw`

coko156 committed:
Update mysqlnd_azure_log.md

Typo fixed

GuuBu reviewed (approved):
will merge and reorganize the doc, single file should be ok.

GuuBu committed:
Merge branch 'master' into add_log
